lib/ansible/plugins/inventory/hcloud.py

```diff
@@ -0,0 +1,204 @@
+          default: []
+          type: list
+          required: false
+        label_selector:
+          description: Populate inventory with instances with this label.
+          default: ""
+          type: str
+          required: false
+"""
+
+EXAMPLES = r"""
+# Minimal example. `HCLOUD_TOKEN` is exposed in environment.
+plugin: hcloud
+
+# Example with locations, types, groups and token
+plugin: hcloud
+token: foobar
+locations:
+  - nbg1
+types:
+  - cx11
+"""
+
+import os
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.module_utils._text import to_native
+from ansible.plugins.inventory import BaseInventoryPlugin
+from ansible.release import __version__
+
+try:
+    from hcloud import hcloud
+except ImportError:
+    raise AnsibleError("The Hetzner Cloud dynamic inventory plugin requires hcloud-python.")
+
+
+class InventoryModule(BaseInventoryPlugin):
+    NAME = "hcloud"
+
+    def _configure_hcloud_client(self):
+        self.api_token = self.get_option("token")
+        if self.api_token is None:
+            self.api_token = os.getenv("HCLOUD_TOKEN")
+            if self.api_token is None:
+                raise AnsibleError(
+                    "Please specify a token, via the option token or via environment variable HCLOUD_TOKEN")
+
+        self.endpoint = os.getenv("HCLOUD_ENDPOINT")
+        if self.endpoint is None:
+            self.endpoint = "https://api.hetzner.cloud/v1"
+
+        self.client = hcloud.Client(token=self.api_token,
+                                    api_endpoint=self.endpoint,
+                                    application_name="ansible-inventory",
+                                    application_version=__version__)
+
+    def _test_hcloud_token(self):
+        try:
+            # We test the API Token against the location API, because this is the API with the smallest result
+            # and not controllable from the customer.
+            self.client.locations.get_all()
+        except hcloud.APIException:
+            raise AnsibleError("Invalid Hetzner Cloud API Token.")
+
+    def _add_groups(self):
+        locations = self.client.locations.get_all()
+        for location in locations:
+            self.inventory.add_group(to_native("location_" + location.name))
+
+        images = self.client.images.get_all(type="system")
+        for image in images:
+            self.inventory.add_group(to_native("image_" + image.os_flavor))
+
+        server_types = self.client.server_types.get_all()
+        for server_type in server_types:
+            self.inventory.add_group(to_native("server_type_" + server_type.name))
+
+    def _get_servers(self):
+        if len(self.get_option("label_selector")) > 0:
+            self.servers = self.client.servers.get_all(label_selector=self.get_option("label_selector"))
+        else:
+            self.servers = self.client.servers.get_all()
+
+    def _filter_servers(self):
```

Comment:
```suggestion
    def _filter_servers(self):
        def server_entry_is_valid(s):
            if s.datacenter.location.name in self.get_option('locations'):
                return True
            if s.server_type.name in self.get_option('types'):
                return True
            if s.image in self.get_option('images'):
                return True
            return False
        return [s for s in self.servers if server_entry_is_valid(s)]
```

Reply:
We have decided to let the things as it is at the moment, because it is just "simple".